File: Data Structures & Algorithms/linked-list-cycle-detection/submission-2.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        if not head or not head.next: return False
        slow = head.next
        fast = head.next.next
        while fast != None and slow != fast:
            if fast.next == None:
                break
            slow = slow.next
            fast = fast.next.next
        return slow == fast

File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-2.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode temp = head;
        int len = 0;

        while (temp != null){
            temp = temp.next;
            len += 1;
        };

        int index = len - n;
        System.out.println(index);
        if (index == 0){
            return head.next;
        }
        var res = head;
        var prev = head;
        while (index > 0){
            prev = head;
            head = head.next;
            index -=1;
        }
        prev.next = head.next;
        return res;
    }
}
